update physics server to contain all methods and implement most of them

Ughuuu opened the pull request:
`register_types`:
Registers and exposes box2d_direct_space_state(not fully impl).

`physics_server_box2d`
Exposes all methods and calls into specific component. Also adds `joint_owner`. For now the idea of joint is that it keeps information on both joint object(about bodies) and on body(and when needed to be created, it should create joint also[not done yet]).

`box2d_space`
Exposes solver_iterations and step_interval.

`box2d_shape`
Adds Edge Shape(works with one way collision too), World Boundary Shape and Separation Ray Shape(not fully impl yet).

`box2d_joint`
For now doesn't do much, just adds values and keeps hold of bodies it should tie.

`box2d_direct_space_state`
Just added class, didn't impl anything yet.

`box2d_direct_body_state`
Exposed methods and implemented a few from body.

`box2d_collision_object`
Added a lot of methods here to be able to call from direct_body_state. Implemented a few of them.

`box2d_body`
Added here new methods that shouldn't be added on box2d_collision_object(only on body, not generic enough).

`box2d_area`
Exposed all methods that come from physics server. Not impl yet.



With this all methods should be exposed form outside.

- Fixes https://github.com/rburing/physics_server_box2d/issues/9
- Fixes https://github.com/rburing/physics_server_box2d/issues/8

Ughuuu committed:
update api and files

fix endlines

lint

update Sconstruct

remove recreate_shape

update

update/fix build

fix crash

run tidy

fix is point, make it bool

fix

upd

Ughuuu commented:
I have thought a lot and I think adding the whole API at once is more beneficial as this way it's easier to do other implementations once the API is decided. Otherwise there are a lot of decisions that remain and might create a lot of conflicts later on. I am still working on the Joints one, but once that is done the API would have all methods expose from physics server 2d at least(not implemented fully, just exposed and calling into the correct place)

Ughuuu commented:
Merging this into my fork and continue work on that. If this remains active I can make a PR with other changes I make there later.

Ughuuu commented:
Updated description of this PR:

First, updates SConstruct to build with folders. Moves files into folders based on what they do to make it easier to organize:
- bodies for rigidbodies, areas, collision object and direct body state
- joints for joints
- servers for the server extension api class
- shapes for the shapes(here i also split the classes into it's own file each)
- spaces for space, direct space state, and callback classes used in direct state space class

Then, updated the PhysicsServerBox2D to override all methods. It also calls into corresponding place(eg. `_joint_set_param` -> `joint->set_param`) similar to how it was for other things that were already working.

## register_types

It registers box2d_direct_space_state and implements raycast and point collision.

## spaces


For intersect ray it uses `b2World->RayCast` and for point it uses `b2World->QueryAABB`. The callback classes are:
- box2d_query_callback for aabb
- box2d_ray_cast_callback for raycast


For except bodies it uses `box2d_space_contact_filter` that is set as `world->SetContactFilter`. This method returns should collide if no body is excepted.

For `call_area_monitor` and `call_monitor` methods from areas I use `world->SetContactListener`. I use `BeginContact` and `EndContact` to call the correct callback from areas. This then calls into the area.


For contacts from space it uses world->GetContactCount and world->GetContactList.

It also adds to step before_step used to do custom gravity and constant forces. Also, world gravity is set to 0 and all gravity is done as custom gravity.


Joints are also created now in space.

## shapes

Aside from moving everything into it's own file,
- added support for EdgeShape, as well as send now `bool one_way, bool is_static` to the `get_transformed_b2Shape` method. The is_static is used for EdgeShape as this is either a `b2EdgeShape` if it's static, or a `b2PolygonShape` if it's dynamic(a rectangle with a small height.
- added support for WorldBoundary(for now it's static only and it's a line).

## servers

Nothing groundbreaking here, added joint_owner and also default area is added to body when created. Also added support to edit default area(if no area is found, the server changes default area.

## joints

Credits to [godot-box2d](https://github.com/briansemrau/godot_box2d) implementation. As per their readme:
- [Distance joint](https://box2d.org/documentation/md__d_1__git_hub_box2d_docs_dynamics.html#autotoc_md85) (Equivalent to GrooveJoint2D or DampedSpringJoint2D)
- [Revolute joint](https://box2d.org/documentation/md__d_1__git_hub_box2d_docs_dynamics.html#autotoc_md86) (PinJoint2D, but with configurable limits and a motor)
- [Prismatic joint](https://box2d.org/documentation/md__d_1__git_hub_box2d_docs_dynamics.html#autotoc_md87) (Similar to GrooveJoint2D, but with fixed rotation)

For pin joint, not using softness. Using the b2RevoluteJoint, which is similar to the godot one but doesn't have softness.

For groove joint, using b2PrismaticJoint.

For damped spring joint, using b2DistanceJoint.

Supports changing already existing joints or joints that aren't created. Most of the data is held on struct's from the class or on the `joint_def`. After it's created, it's a `b2Joint` that can be then cast into specific class if needed to change something from it. Only property that requires recreating the joint is `collideConnected` as it can't be changed from `joint` only from `joint_def`.

The joint also holds a reference to a space that it needs to call to remove itself or recreate itself.

## bodies - areas

Areas have implemented gravity and linear/angular damp. Areas have a vector of bodies that are updated from b2World collision callback. Bodies also have a vector of areas they have. Areas need to have a list of bodies in case their priority is updated, the areas from bodies need to be resorted and all properties recalculated. Bodies need to have areas in order to calculate all the properties(gravity, linear/angular damp).

Another important thing is areas are now b2_dynamicBody. This is needed because otherwise they cannot collide with other areas or with static or kinematic boddies.

## bodies - body

Bodies have a hash with joints they have. This is needed to know what joints to invalidate when destroying a body.

Also, updated most of the properties from body that already exists and added checks to not change properties in case they are same(eg. 
```
	if (mode == p_mode) {
		return;
	}
```
Same for other places with these kind of checks.

Also added support for fixed rotation: `BODY_MODE_RIGID_LINEAR` is fixed rotation and `BODY_MODE_RIGID` is regular dynamic object.

Also added support for `bullet` which is box2d fast moving objects.

## bodies - collision object

Added support to sort based on priority, collision exceptions, areas, collision filter, constant forces, physics material, damping and mass.

For Collision exception and collision fitlering, this is done from b2World filtering/contact callback.

For constant forces and areas/(custom gravity), this is done through the before_step callback.

For damping and gravity, uses areas and default areas to compute the total gravity and total damp.

Box2dDirectBodyState calls mostly into body methods.

## misc

If there are invalid polygons, the client crashes(because of 
`box2d/b2_common.h > #define b2Assert(A) assert(A)`
If this is changed instead to 
`#define b2Assert(A) ((void)(x))`
then it doesn't crash, but if you have a very small square, it creates a 1x1 square for eg. I already added a lot of checks for shapes, but more are needed in case of polygons. Not sure how this could be fixed, either a `sed` command that changes that, or a fork repo instead of a subrepo, or maybe some redefine?

As for the [Godot-Physics-Tests](https://github.com/fabriceci/Godot-Physics-Tests), I ran them and currently have 77/150 passing tests.

rburing commented on src/b2_user_settings.h:
```diff
@@ -30,12 +29,14 @@ struct B2_API b2BodyUserData {
 
 	int shape_idx;
 	int box2d_fixture_idx;
+	int type;
```
This seems to be unused; I would think it's unnecessary to add here.

Ughuuu replied:
fixed

rburing replied:
Looks like it came back. After that fix (and squash) this looks good to merge!

Ughuuu commented:
There is a conflict now because on actions branch I ran lint and that changed 1 file. Will fix conflict.

Ughuuu commented:
Ready to be merged by me. Might add more if you don't merge it, if you do will open smaller pr's with functionality. As is it adds a lot though and it is stable.

rburing commented:
Really great work!

Could you please squash the commits?

rburing commented on src/bodies/box2d_area.h:
```diff
@@ -0,0 +1,73 @@
+		PhysicsServer2D::AreaSpaceOverrideMode override_mode = PhysicsServer2D::AreaSpaceOverrideMode::AREA_SPACE_OVERRIDE_DISABLED;
+		real_t gravity = 9.8;
+		b2Vec2 vector = b2Vec2(0, 1);
+		real_t is_point = false;
```
This should be a `bool` (the actions you added caught this!)

Ughuuu replied:
fixed.(strange that it is a warning and wasn't an error)

Ughuuu commented:
Squashed

Ughuuu committed:
fix shapes

fix bug

add impl for separation ray also

Ughuuu commented:
Added 1 more commit that fixes all of the shapes:

For convexPolygon, it splits into 8 polygons if possible. Based on link you provided from other repo.
For concavePolygon, if it's static, creates a simple ChainShape, if not, creates multiple squares in the shape of each segment.
For boundary shape, fixes the case where the boundary is dynamic so it can move.
For separation ray, an edge shape for now.

Ughuuu committed:
add missing filtering for objects

Ughuuu commented:
Also fixed filtering for objects(when I added custom ShouldCollide extension, it's not done automatically, the mask and group check bits). Now it works. Test count is now at 78/150.

rburing reviewed (approved):
Thanks a lot for all this work! I hope to be helping out more later.
